paddle/operators/momentum_op.h

```diff
@@ -44,11 +47,28 @@ class MomentumOpKernel : public framework::OpKernel<T> {
     auto g = framework::EigenVector<T>::Flatten(*grad);
     auto* lr = learning_rate->data<T>();
 
+    T local_lr = lr[0];
+    if (use_local_lr) {
+      Eigen::TensorFixedSize<T, Eigen::Sizes<>, Eigen::RowMajor,
+                             Eigen::DenseIndex>
+          p_norm = p.square().sum().sqrt();
```

Comment:
Assigning to tensor may cause Eigen to do evaluation here, which may cause performance issue.

See: https://bitbucket.org/eigen/eigen/src/default/unsupported/Eigen/CXX11/src/Tensor/README.md?fileviewer=file-view-default#markdown-header-controlling-when-expression-are-evaluated

Reply:
Thanks for the careful review!  if "auto" is used instead of "Tensor",   we can not perform p_norm > 0 and g_norm > 0 check due to lazy evaluation, there will be compilation  error.

Reply:
Can add a epsilon to avoid zero-division.

Reply:
It's weird if I use "auto" to calculate "param_out" as below, some elements in c++ output will be "-nan" and can't match python testcase output.  yet the g_norm/p_norm/local_lr  output are the same in both python and c++. 

      if (use_local_lr) {
        auto p_norm = p.square().sum().sqrt();
        auto g_norm = g.square().sum().sqrt();
        auto local_lr = lr[0] * local_gw_ratio * (p_norm + epsilon) /
                        (g_norm + weight_decay * p_norm + epsilon);
        o = p - local_lr * g;
        std::cout << o;
      } else {
        o = p - lr[0] * g;
      }

here is python output for param_out
[0.69548035 0.28497267 0.22662182 0.5497418  0.718321   0.4220689   0.9801117  0.68425083]

here is c++ output for param_out
  [ 0.69548   0.286139 -2.00666e+21  0.551315 -4.41094e+34  0.423106  0.980764   0.68483]
     
if I set the matrix size of weight and grad very small  such as (4,5), the test_sgd_op.py unit test can pass, 

 I'm afraid it's due to the expression is a little long and lead to such error. if I use lr[0] to replace local_lr, such param_out discrepancy won't occur.

Any suggestion or comments? Thanks.

Reply:
if I fallback to use tensor to calculate local_lr as below,  param_out in both c++ and python can match and test case can pass.
    T local_lr = lr[0];
      if (use_local_lr) {
        Eigen::TensorFixedSize<T, Eigen::Sizes<>, Eigen::RowMajor,
                       Eigen::DenseIndex> p_norm = p.square().sum().sqrt();
        Eigen::TensorFixedSize<T, Eigen::Sizes<>, Eigen::RowMajor,
                       Eigen::DenseIndex> g_norm = g.square().sum().sqrt();
        local_lr = lr[0] * local_gw_ratio * (p_norm(0) + epsilon) /
                        (g_norm(0) + weight_decay * p_norm(0) + epsilon);
      }
      o = p - local_lr * g;
      std::cout << o;

Reply:
If I use "auto" to calculate  local_lr an param_out as below, local_lr can match in c++ print and python unit test case. 
But for  "o = p - local_lr * g",  some elements in "o" are -nan  in c++ print.  I add  eval for local_lr such as "o = p - local_lr.eval() * g", some element in "o" are very strange negative values as mentioned in comments above. 
whether or not I add epsilon, the python test case " TestSGDLARSOp" and "TestMomentumOp3" failed.
So I fall back to use Eigen tensor to calculate local_lr and removed epsilon.  python unit test can pass.  Any comments or suggestion? Thanks!

 if (use_local_lr) {
    auto p_norm = p.square().sum().sqrt();
    auto g_norm = g.square().sum().sqrt();
    auto local_lr = lr[0] * local_gw_ratio * (p_norm + epsilon) /
                    (g_norm + weight_decay * p_norm + epsilon);
    o = p - local_lr * g;
    std::cout << o;
  } else {
    o = p - lr[0] * g;
  }